File: 백준/Silver/2583. 영역 구하기/영역 구하기.cc
#include <bits/stdc++.h>

using namespace std;

int m, n, k;

int Map[101][101];
int visited[101][101];
int dx[] = { 0,0,-1,1 };
int dy[] = { 1,-1, 0, 0 };

typedef struct node {
	int x;
	int y;
	int cnt;
};

queue<node> q;
vector<int> v;

int bfs(int i, int j)
{
	q.push(node{ i,j,1 });
	visited[i][j] = 1;
	
	int val = 0;

	while (!q.empty())
	{
		int cx = q.front().x;
		int cy = q.front().y;
		int cnt = q.front().cnt;

		q.pop();

		val++;

		for (int i = 0; i < 4; i++)
		{
			int nx = cx + dx[i];
			int ny = cy + dy[i];

			if (nx < 0 || ny < 0 || nx >= m || ny >= n || visited[nx][ny] || Map[nx][ny])
				continue;
			else
			{
				q.push(node{ nx,ny,cnt + 1 });
				visited[nx][ny] = 1;
			}
		}
	}

	return val;
}

int main()
{
	cin >> m >> n >> k;

	for (int i = 0; i < k; i++)
	{
		int x1, y1, x2, y2;
		cin >> y1 >> x1 >> y2 >> x2;

		for (int i = x1; i < x2; i++)
		{
			for (int j = y1; j < y2; j++)
			{
				Map[i][j] = 1;
			}
		}
	}

	for (int i = 0; i < m; i++)
	{
		for (int j = 0; j < n; j++)
		{
			if (Map[i][j] != 1 && !visited[i][j])
			{
				int result = bfs(i,j);
				v.push_back(result);
			}
		}
	}

	sort(v.begin(), v.end());

	cout << v.size() << "\n";

	for (int i = 0; i < v.size(); i++)
	{
		cout << v[i] << " ";
	}

	return 0;
}

File: 백준/Silver/2583. 영역 구하기/영역 구하기.java
import java.util.*;
import java.io.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    static int n, m, k, result;
    static int[][] map, visited;
    static ArrayList<Integer> list = new ArrayList<>();

    static int[] dx = {0,0,-1,1};
    static int[] dy = {1,-1,0,0};

    public static void main(String[] args) throws Exception {

        st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());

        map = new int[n][m];
        visited = new int[n][m];

        for (int i = 0 ; i < k ; i ++) {
            st = new StringTokenizer(br.readLine());

            int y1 = Integer.parseInt(st.nextToken());
            int x1 = Integer.parseInt(st.nextToken());

            int y2 = Integer.parseInt(st.nextToken());
            int x2 = Integer.parseInt(st.nextToken());

            for (int X = x1 ; X < x2 ; X ++) {
                for (int Y = y1 ; Y < y2 ; Y ++) {
                    map[X][Y] = 1;
                }
            }
        }

        for (int i = 0 ; i < n ; i ++) {
            for (int j = 0 ; j < m ; j ++) {
                if (visited[i][j] == 0 && map[i][j] == 0) {
                    result ++;
                    bfs(i, j);
                }
            }
        }

        Collections.sort(list);

        sb.append(result).append('\n');

        for (int v : list) {
            sb.append(v).append(' ');
        }

        System.out.println(sb);
    }

    static class Node {
        int x;
        int y;

        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static void bfs(int x, int y) {
        ArrayDeque<Node> q = new ArrayDeque<>();
        visited[x][y] = result;
        q.add(new Node(x, y));
        int count = 0;

        while (!q.isEmpty()) {
            Node cur = q.poll();
            int cx = cur.x;
            int cy = cur.y;
            count ++;

            for (int i = 0 ; i < 4; i ++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if (nx < 0 || ny < 0 || nx >= n || ny >= m) continue;

                if (visited[nx][ny] != 0 || map[nx][ny] == 1) continue;

                q.add(new Node(nx, ny));
                visited[nx][ny] = result;
            }
        }

        list.add(count);
    }
}